### PR TITLE
Push all to remote

### PR DESCRIPTION
**Pull request by technoweenie**
This uses the `lfs.ScanRefsMode` scan mode when pushing if the `--all` flag is given. This lets you confirm that everything referenced in the given ref has been pushed to the given remote.

### CODE REVIEW

File: commands/command_push.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ var (
 	pushDryRun       = false
 	pushDeleteBranch = "(delete)"
 	pushObjectIDs    = false
+	pushAllToRemote  = false
```
Comment (ttaylorr) [inline commands/command_push.go]: Do you think the `ToRemote` part of this identifier could be dropped?

**Comment by ttaylorr** [comment]
Other than my one comment, looks good to me :+1:

**Comment by sinbad** [comment]
You need to update the man page too. :)

I like the concept, but I think its interpretation of `--all` is potentially misleading. It's "all at one ref" not "all binaries ever", which is not symmetrical with `git lfs fetch --all` which downloads all referenced binaries ever (for backup). I'm concerned that people will think they can perform a complete migration of their LFS files using `fetch --all` then `push --all` and they'd be done - but they won't be because as currently implemented it would actually implicitly mean `push --all HEAD` only.

I think `git lfs push --all` with no ref parameter should mean `push all binaries ever` i.e. using ScanAllMode. I think the behaviour in this PR i.e. ScanRefsMode should be reserved for *either* a different option (`--all-ref`?) in which case ref args should be disallowed, or alternatively ScanRefsMode should only be used when a ref is given *explicitly* on the command line, not implied HEAD.

**Comment by sinbad** [comment]
Also the non-migratory version of `--all` might need a step like `prePushCheckForMissingObjects` to cope with cases where the person pushing doesn't have the entire LFS state.

File: commands/command_push.go (hunk 1 above)
Comment (technoweenie) [inline commands/command_push.go]: I wanted to distinguish this from `fetch --all`, which downloads EVERYTHING. Though `pushAllForRef` would be a more accurate var name I think.

**Comment by technoweenie** [comment]
I totally updated the man page, but forgot to push. Or I faked the commit times :trollface: 

`fetch` symmetry is a better idea. :+1:

```bash
# PUSH EVERYTHING
$ git lfs push --all origin

# just master branch
$ git lfs push --all origin master

# multiple refs
$ git lfs push --all origin master feature/1 v1.0
```

I'll add a test case with missing objects too. Good catch.

**Comment by sinbad** [comment]
> I totally updated the man page, but forgot to push. Or I faked the commit times :troll face:

I believe you ;)

> `# PUSH EVERYTHING`...

:+1:

**Comment by technoweenie** [comment]
The behavior matches https://github.com/github/git-lfs/pull/646#issuecomment-138901295 now. Still a couple issues:

1. Probably want to use `lfs.ScanRefsToChan()` like `fetch --all` does. It has its own spinner while it's scanning the entire history.
2. `prePushCheckForMissingObjects()` outputs a separate progress output. 

        $ git lfs push --all origin master 
        (1 of 1 files) 24.65 KB / 24.65 KB                                                                                                                                                                                                                                                                                                                                         
        (0 of 3 files, 2 skipped) 0 B / 1.45 MB, 1.43 MB skipped

**Comment by technoweenie** [comment]
Basically copied a bunch of spinner code from `fetchAll()` to get this:

```
$ git lfs push --all origin
Scanning for all objects ever referenced...
✔ 3 objects found                                                                                                                                                                                                                                                                                                                                                           
Pushing objects...
(1 of 1 files) 24.65 KB / 24.65 KB                                                                                                                                                                                                                                                                                                                                         
(0 of 3 files, 2 skipped) 0 B / 1.45 MB, 1.43 MB skipped     
```

**Comment by technoweenie** [comment]
I removed the first progress meter by setting `dryRun` to true.

```
$ git lfs push --all origin
Scanning for all objects ever referenced...
✔ 3 objects found                                                                                                                                                                                                                                                                                                                                                           
Pushing objects...
(0 of 3 files, 3 skipped) 0 B / 1.45 MB, 1.45 MB skipped   
```

It still writes the progress log if `GIT_LFS_PROGRESS` is passed, though.

File: docs/man/git-lfs-push.1.ronn [hunk 2]
```diff
@@ -3,25 +3,28 @@ git-lfs-push(1) -- Push queued large files to the Git LFS endpoint
...
+of the remote.
 
 ## OPTIONS
 
+* `--dry-run`:
+    Print the files that would be pushed, without actually pushing them.
+
+* `--all`:
+    This pushes all objects for the given ref or current branch, regardless of
+    whether they are referenced in the local clone of the remote.
```
Comment (sinbad) [inline docs/man/git-lfs-push.1.ronn]: Should really mention the difference between providing a ref on the command line & not when `--all` is provided.

**Comment by sinbad** [comment]
> I removed the first progress meter by setting dryRun to true.

Wow, that's entirely *not* what I would expect `dryRun` to do. I think that option should be renamed to `suppressProgress` instead to avoid confusion in future, because that's what it actually does - looks to me like `dryRun=true` does everything `dryRun=false` does apart from the progress output. I can imagine a time when someone fixes it to *actually* be a dry run option (and thus not do anything except local scanning) and for that to break this code.

Other than that and my man comment, looks good to me! :+1:

File: docs/man/git-lfs-push.1.ronn (hunk 2 above)
Comment (technoweenie) [inline docs/man/git-lfs-push.1.ronn]: "This pushes all objects to the remote that are referenced by any reachable commit. If no ref is provided, then all refs are pushed."

**Comment by technoweenie** [comment]
I agree about `dryRun`, but I want to update that in a separate PR.

**Pull request merged**